pkg/node/peers/score_selector.go

```diff
@@ -0,0 +1,166 @@
+package peers
+
+import (
+	"container/heap"
+	"fmt"
+
+	"golang.org/x/exp/rand"
+
+	"github.com/wavesplatform/gowaves/pkg/p2p/peer"
+	"github.com/wavesplatform/gowaves/pkg/proto"
+)
+
+type group struct {
+	score *proto.Score
+	peers []peer.ID
+	index int
+}
+
+type groupsHeap struct {
+	groups []*group
+}
+
+func newGroupsHeap() *groupsHeap {
+	return &groupsHeap{groups: make([]*group, 0)}
+}
+
+func (h *groupsHeap) Len() int {
+	return len(h.groups)
+}
+
+func (h *groupsHeap) Less(i, j int) bool {
+	// Pop selects the group with the largest size
+	return len(h.groups[i].peers) > len(h.groups[j].peers)
+}
+
+func (h *groupsHeap) Swap(i, j int) {
+	h.groups[i], h.groups[j] = h.groups[j], h.groups[i]
+	h.groups[i].index = i
+	h.groups[j].index = j
+}
+
+func (h *groupsHeap) Push(x any) {
+	g, ok := x.(*group)
+	if !ok {
+		panic(fmt.Sprintf("groupsHeap: invalid element type: expected (*group), got (%T)", x))
+	}
+	g.index = len(h.groups)
+	h.groups = append(h.groups, g)
+}
+
+func (h *groupsHeap) Pop() any {
+	n := len(h.groups)
+	g := h.groups[n-1]
+	h.groups[n-1] = nil
+	g.index = -1 // for safety
+	h.groups = h.groups[:n-1]
+	return g
+}
+
+type (
+	peerID   string
+	scoreKey string
+)
+
+type scoreSelector struct {
+	scoreKeyToGroup map[scoreKey]*group // key made of score's string points to the group of peers
+	peerToScoreKey  map[peerID]scoreKey
+	groups          *groupsHeap
+}
+
+func newScoreSelector() *scoreSelector {
+	return &scoreSelector{
+		scoreKeyToGroup: make(map[scoreKey]*group),
+		peerToScoreKey:  make(map[peerID]scoreKey),
+		groups:          newGroupsHeap(),
+	}
+}
+
+func (s *scoreSelector) push(p peer.ID, score *proto.Score) {
+	sk := scoreKey(score.String())
+	pid := peerID(p.String())
+	if prevScoreKey, ok := s.peerToScoreKey[pid]; ok {
+		// The peer was added before
+		if sk == prevScoreKey { // Do nothing, if the score of the peer hasn't changed.
+			return
+		}
+		// Remove the peer from the previous score group
+		s.remove(prevScoreKey, p)
+		s.append(sk, score, p)
+	} else {
+		s.append(sk, score, p)
+	}
+}
+
+func (s *scoreSelector) delete(p peer.ID) {
+	pid := peerID(p.String())
+	if sk, ok := s.peerToScoreKey[pid]; ok {
+		s.remove(sk, p)
+		delete(s.peerToScoreKey, pid)
```

Comment:
Why do you not put `delete(s.peerToScoreKey, pid)` inside `s.remove` method?

Reply:
Function `remove` is used to move peer from one group to another. Not only to delete peer. While moving no need to remove peer itself. In turn, `delete` is called then peer disconnects and can be removed from both group and peer references.